File: expectanim/src/main/java/com/github/florent37/expectanim/core/Expectations.java
package com.github.florent37.expectanim.core;

import android.view.View;

import com.github.florent37.expectanim.GravityIntDef;
import com.github.florent37.expectanim.GravityScaleHorizontalIntDef;
import com.github.florent37.expectanim.GravityScaleVerticalIntDef;
import com.github.florent37.expectanim.core.alpha.AlphaAnimExpectation;
import com.github.florent37.expectanim.core.alpha.AlphaAnimExpectationValue;
import com.github.florent37.expectanim.core.anim3d.CameraDistanceExpectation;
import com.github.florent37.expectanim.core.anim3d.CameraDistanceExpectationValue;
import com.github.florent37.expectanim.core.custom.CustomAnimExpectation;
import com.github.florent37.expectanim.core.custom.TextColorAnimExpectation;
import com.github.florent37.expectanim.core.custom.ViewBackgroundAlphaAnimExpectation;
import com.github.florent37.expectanim.core.position.PositionAnimExpectation;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationAboveOf;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationAlignBottom;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationAlignLeft;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationAlignRight;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationAlignTop;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationBelowOf;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationBottomOfParent;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationCenterBetweenViewAndParent;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationCenterBetweenViews;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationCenterInParent;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationLeftOf;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationLeftOfParent;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationOriginal;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationOutOfScreen;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationRightOf;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationRightOfParent;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationSameCenterAs;
import com.github.florent37.expectanim.core.position.PositionAnimExpectationTopOfParent;
import com.github.florent37.expectanim.core.rotation.RotationExpectation;
import com.github.florent37.expectanim.core.rotation.RotationExpectationOriginal;
import com.github.florent37.expectanim.core.rotation.RotationExpectationValue;
import com.github.florent37.expectanim.core.rotation.RotationFlipExpectationValue;
import com.github.florent37.expectanim.core.scale.ScaleAnimExpectation;
import com.github.florent37.expectanim.core.scale.ScaleAnimExpectationHeight;
import com.github.florent37.expectanim.core.scale.ScaleAnimExpectationOriginalScale;
import com.github.florent37.expectanim.core.scale.ScaleAnimExpectationSameHeightAs;
import com.github.florent37.expectanim.core.scale.ScaleAnimExpectationSameScaleAs;
import com.github.florent37.expectanim.core.scale.ScaleAnimExpectationSameWidthAs;
import com.github.florent37.expectanim.core.scale.ScaleAnimExpectationValues;
import com.github.florent37.expectanim.core.scale.ScaleAnimExpectationWidth;

import java.util.ArrayList;
import java.util.List;

public class Expectations {

    private List<AnimExpectation> expectations = new ArrayList<>();

    //region position

    public PositionAnimExpectation toRightOf(View view) {
        final PositionAnimExpectation expect = new PositionAnimExpectationRightOf(view);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation toLeftOf(View view) {
        final PositionAnimExpectation expect = new PositionAnimExpectationLeftOf(view);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation belowOf(View view) {
        final PositionAnimExpectation expect = new PositionAnimExpectationBelowOf(view);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation aboveOf(View view) {
        final PositionAnimExpectationAboveOf expect = new PositionAnimExpectationAboveOf(view);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation atItsOriginalPosition() {
        final PositionAnimExpectation expect = new PositionAnimExpectationOriginal();
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation sameCenterAs(View view, boolean horizontal, boolean vertical) {
        final PositionAnimExpectationSameCenterAs expect = new PositionAnimExpectationSameCenterAs(view, horizontal, vertical);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation sameCenterHorizontalAs(View view) {
        final PositionAnimExpectation expect = sameCenterAs(view, true, false);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation sameCenterVerticalAs(View view) {
        final PositionAnimExpectation expect = sameCenterAs(view, false, true);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation centerInParent(boolean horizontal, boolean vertical) {
        final PositionAnimExpectation expect = new PositionAnimExpectationCenterInParent(horizontal, vertical);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation centerVerticalInParent() {
        final PositionAnimExpectation expect = centerInParent(false, true);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation centerHorizontalInParent() {
        final PositionAnimExpectation expect = centerInParent(true, false);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation outOfScreen(@GravityIntDef int... gravities) {
        final PositionAnimExpectationOutOfScreen expect = new PositionAnimExpectationOutOfScreen(gravities);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation centerBetweenViews(View view1, View view2, boolean horizontal, boolean vertical) {
        final PositionAnimExpectation expect = new PositionAnimExpectationCenterBetweenViews(view1, view2, horizontal, vertical);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation topOfParent() {
        final PositionAnimExpectationTopOfParent expect = new PositionAnimExpectationTopOfParent();
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation rightOfParent() {
        final PositionAnimExpectationRightOfParent expect = new PositionAnimExpectationRightOfParent();
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation bottomOfParent() {
        final PositionAnimExpectationBottomOfParent expect = new PositionAnimExpectationBottomOfParent();
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation leftOfParent() {
        final PositionAnimExpectationLeftOfParent expect = new PositionAnimExpectationLeftOfParent();
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation centerBetweenViewAndParent(View otherView, boolean horizontal, boolean vertical, boolean toBeOnRight, boolean toBeOnBottom) {
        final PositionAnimExpectationCenterBetweenViewAndParent expect = new PositionAnimExpectationCenterBetweenViewAndParent(otherView, horizontal, vertical, toBeOnRight, toBeOnBottom);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation alignBottom(View otherView) {
        final PositionAnimExpectationAlignBottom expect = new PositionAnimExpectationAlignBottom(otherView);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation alignTop(View otherView) {
        final PositionAnimExpectationAlignTop expect = new PositionAnimExpectationAlignTop(otherView);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation alignLeft(View otherView) {
        final PositionAnimExpectationAlignLeft expect = new PositionAnimExpectationAlignLeft(otherView);
        expectations.add(expect);
        return expect;
    }

    public PositionAnimExpectation alignRight(View otherView) {
        final PositionAnimExpectationAlignRight expect = new PositionAnimExpectationAlignRight(otherView);
        expectations.add(expect);
        return expect;
    }

    //endregion

    //region alpha

    public AlphaAnimExpectation alpha(float alpha) {
        final AlphaAnimExpectationValue expect = new AlphaAnimExpectationValue(alpha);
        expectations.add(expect);
        return expect;
    }

    public AlphaAnimExpectation sameAlphaAs(View otherView) {
        final AlphaAnimExpectationValue expect = new AlphaAnimExpectationValue(otherView.getAlpha());
        expectations.add(expect);
        return expect;
    }

    public AlphaAnimExpectation visible() {
        final AlphaAnimExpectationValue expect = new AlphaAnimExpectationValue(1f);
        expectations.add(expect);
        return expect;
    }

    public AlphaAnimExpectation invisible() {
        final AlphaAnimExpectationValue expect = new AlphaAnimExpectationValue(0f);
        expectations.add(expect);
        return expect;
    }

    //endregion

    //region scale


    public ScaleAnimExpectation atItsOriginalScale() {
        final ScaleAnimExpectationOriginalScale expect = new ScaleAnimExpectationOriginalScale();
        expectations.add(expect);
        return expect;
    }

    public ScaleAnimExpectation scale(float scaleX, float scaleY, @GravityScaleHorizontalIntDef int gravityHorizontal, @GravityScaleVerticalIntDef int gravityVertical) {
        final ScaleAnimExpectationValues expect = new ScaleAnimExpectationValues(scaleX, scaleY, gravityHorizontal, gravityVertical);
        expectations.add(expect);
        return expect;
    }

    public ScaleAnimExpectation scale(float scaleX, float scaleY) {
        final ScaleAnimExpectationValues expect = new ScaleAnimExpectationValues(scaleX, scaleY, null, null);
        expectations.add(expect);
        return expect;
    }

    public ScaleAnimExpectation height(int height, @GravityScaleHorizontalIntDef int gravityHorizontal, @GravityScaleVerticalIntDef int gravityVertical) {
        final ScaleAnimExpectationHeight expect = new ScaleAnimExpectationHeight(height, gravityHorizontal, gravityVertical);
        expectations.add(expect);
        return expect;
    }

    public ScaleAnimExpectation height(int height) {
        final ScaleAnimExpectationHeight expect = new ScaleAnimExpectationHeight(height, null, null);
        expectations.add(expect);
        return expect;
    }

    public ScaleAnimExpectation width(int width, @GravityScaleHorizontalIntDef int gravityHorizontal, @GravityScaleVerticalIntDef int gravityVertical) {
        final ScaleAnimExpectationWidth expect = new ScaleAnimExpectationWidth(width, gravityHorizontal, gravityVertical);
        expectations.add(expect);
        return expect;
    }

    public ScaleAnimExpectation width(int width) {
        final ScaleAnimExpectationWidth expect = new ScaleAnimExpectationWidth(width, null, null);
        expectations.add(expect);
        return expect;
    }

    public ScaleAnimExpectation sameScaleAs(View otherView) {
        final ScaleAnimExpectationSameScaleAs expect = new ScaleAnimExpectationSameScaleAs(otherView);
        expectations.add(expect);
        return expect;
    }

    public ScaleAnimExpectation sameWidthAs(View otherView) {
        final ScaleAnimExpectationSameWidthAs expect = new ScaleAnimExpectationSameWidthAs(otherView, null, null);
        expectations.add(expect);
        return expect;
    }

    public ScaleAnimExpectation sameHeightAs(View otherView) {
        final ScaleAnimExpectationSameHeightAs expect = new ScaleAnimExpectationSameHeightAs(otherView, null, null);
        expectations.add(expect);
        return expect;
    }

    //endregion

    //region custom

    public CustomAnimExpectation toHaveTextColor(int textColor) {
        final TextColorAnimExpectation expect = new TextColorAnimExpectation(textColor);
        expectations.add(expect);
        return expect;
    }

    public CustomAnimExpectation toHaveBackgroundAlpha(float alpha) {
        final ViewBackgroundAlphaAnimExpectation expect = new ViewBackgroundAlphaAnimExpectation(alpha);
        expectations.add(expect);
        return expect;
    }


    //endregion

    //region rotation

    public RotationExpectation rotated(float rotation) {
        final RotationExpectationValue expect = new RotationExpectationValue(rotation);
        expectations.add(expect);
        return expect;
    }

    public CameraDistanceExpectation withCameraDistance(float cameraDistance) {
        final CameraDistanceExpectationValue expect = new CameraDistanceExpectationValue(cameraDistance);
        expectations.add(expect);
        return expect;
    }

    public RotationExpectation flippedHorizontally() {
        final RotationFlipExpectationValue expect = new RotationFlipExpectationValue(0f, 180f);
        expectations.add(expect);
        return expect;
    }

    public RotationExpectation flippedVertically() {
        final RotationFlipExpectationValue expect = new RotationFlipExpectationValue(180f, 0f);
        expectations.add(expect);
        return expect;
    }

    public RotationExpectation flippedHorizontallyAndVertically() {
        final RotationFlipExpectationValue expect = new RotationFlipExpectationValue(180f, 180f);
        expectations.add(expect);
        return expect;
    }

    public RotationExpectation vertical(boolean bottomOfViewAtLeft) {
        final RotationExpectation expect;
        if (bottomOfViewAtLeft) {
            expect = new RotationExpectationValue(90);
        } else {
            expect = new RotationExpectationValue(270);
        }
        expectations.add(expect);
        return expect;
    }

    public RotationExpectation atItsOriginalRotation() {
        final RotationExpectationOriginal expect = new RotationExpectationOriginal();
        expectations.add(expect);
        return expect;
    }

    //endregion


    public List<AnimExpectation> getExpectations() {
        return expectations;
    }
}

File: expectanim/src/main/java/com/github/florent37/expectanim/core/Expectations.kt
package com.github.florent37.expectanim.core

import android.view.View
import com.github.florent37.expectanim.GravityIntDef
import com.github.florent37.expectanim.GravityScaleHorizontalIntDef
import com.github.florent37.expectanim.GravityScaleVerticalIntDef
import com.github.florent37.expectanim.core.alpha.AlphaAnimExpectation
import com.github.florent37.expectanim.core.alpha.AlphaAnimExpectationValue
import com.github.florent37.expectanim.core.custom.CustomAnimExpectation
import com.github.florent37.expectanim.core.custom.TextColorAnimExpectation
import com.github.florent37.expectanim.core.custom.ViewBackgroundAlphaAnimExpectation
import com.github.florent37.expectanim.core.position.*
import com.github.florent37.expectanim.core.rotation.RotationExpectation
import com.github.florent37.expectanim.core.rotation.RotationExpectationOriginal
import com.github.florent37.expectanim.core.rotation.RotationExpectationValue
import com.github.florent37.expectanim.core.rotation.RotationFlipExpectationValue
import com.github.florent37.expectanim.core.scale.*

class Expectations {

    internal val expectations = mutableListOf<AnimExpectation>()

    fun rightOf(view: View, marginDp: Float? = null, margin: Float? = null): PositionAnimExpectation {
        return PositionAnimExpectationRightOf(view).apply {
            expectations.add(this)
            this.marginDp = marginDp
            this.margin = margin
        }
    }

    fun leftOf(view: View, marginDp: Float? = null, margin: Float? = null): PositionAnimExpectation {
        return PositionAnimExpectationLeftOf(view).apply {
            expectations.add(this)
            this.marginDp = marginDp
            this.margin = margin
        }
    }

    fun belowOf(view: View, marginDp: Float? = null, margin: Float? = null): PositionAnimExpectation {
        return PositionAnimExpectationBelowOf(view).apply {
            expectations.add(this)
            this.marginDp = marginDp
            this.margin = margin
        }
    }

    fun aboveOf(view: View, marginDp: Float? = null, margin: Float? = null): PositionAnimExpectation {
        return PositionAnimExpectationAboveOf(view).apply {
            expectations.add(this)
            this.marginDp = marginDp
            this.margin = margin
        }
    }

    fun originalPosition(): PositionAnimExpectation {
        return PositionAnimExpectationOriginal().apply {
            expectations.add(this)
        }
    }

    fun sameCenterAs(view: View, horizontal: Boolean = false, vertical: Boolean = false): PositionAnimExpectation {
        return PositionAnimExpectationSameCenterAs(view, horizontal, vertical).apply {
            expectations.add(this)
        }
    }

    fun sameCenterHorizontalAs(view: View): PositionAnimExpectation {
        return sameCenterAs(view, true, false).apply {
            expectations.add(this)
        }
    }

    fun sameCenterVerticalAs(view: View): PositionAnimExpectation {
        return sameCenterAs(view, false, true).apply {
            expectations.add(this)
        }
    }

    fun centerInParent(horizontal: Boolean, vertical: Boolean): PositionAnimExpectation {
        return PositionAnimExpectationCenterInParent(horizontal, vertical).apply {
            expectations.add(this)
        }
    }

    fun centerVerticalInParent(): PositionAnimExpectation {
        return centerInParent(false, true).apply {
            expectations.add(this)
        }
    }

    fun centerHorizontalInParent(): PositionAnimExpectation {
        return centerInParent(true, false).apply {
            expectations.add(this)
        }
    }

    fun outOfScreen(@GravityIntDef vararg gravities: Int): PositionAnimExpectation {
        return PositionAnimExpectationOutOfScreen(gravities).apply {
            expectations.add(this)
        }
    }

    fun centerBetweenViews(view1: View, view2: View, horizontal: Boolean = false, vertical: Boolean = false): PositionAnimExpectation {
        return PositionAnimExpectationCenterBetweenViews(view1, view2, horizontal, vertical).apply {
            expectations.add(this)
        }
    }

    fun topOfHisParent(marginDp: Float? = null, margin: Float? = null): PositionAnimExpectation {
        return PositionAnimExpectationTopOfParent().apply {
            expectations.add(this)
            this.marginDp = marginDp
            this.margin = margin
        }
    }

    fun rightOfHisParent(marginDp: Float? = null, margin: Float? = null): PositionAnimExpectation {
        return PositionAnimExpectationRightOfParent().apply {
            expectations.add(this)
            this.marginDp = marginDp
            this.margin = margin
        }
    }

    fun bottomOfHisParent(marginDp: Float? = null, margin: Float? = null): PositionAnimExpectation {
        return PositionAnimExpectationBottomOfParent().apply {
            expectations.add(this)
            this.marginDp = marginDp
            this.margin = margin
        }
    }

    fun leftOfHisParent(marginDp: Float? = null, margin: Float? = null): PositionAnimExpectation {
        return PositionAnimExpectationLeftOfParent().apply {
            expectations.add(this)
            this.marginDp = marginDp
            this.margin = margin
        }
    }

    fun centerBetweenViewAndHisParent(otherView: View, horizontal: Boolean, vertical: Boolean, toBeOnRight: Boolean, toBeOnBottom: Boolean, marginDp: Float? = null, margin: Float? = null): PositionAnimExpectation {
        return PositionAnimExpectationCenterBetweenViewAndParent(otherView, horizontal, vertical, toBeOnRight, toBeOnBottom).apply {
            expectations.add(this)
            this.marginDp = marginDp
            this.margin = margin
        }
    }

    fun alignBottom(otherView: View, marginDp: Float? = null, margin: Float? = null): PositionAnimExpectation {
        return PositionAnimExpectationAlignBottom(otherView).apply {
            expectations.add(this)
            this.marginDp = marginDp
            this.margin = margin
        }
    }

    fun alignTop(otherView: View, marginDp: Float? = null, margin: Float? = null): PositionAnimExpectation {
        return PositionAnimExpectationAlignTop(otherView).apply {
            expectations.add(this)
            this.marginDp = marginDp
            this.margin = margin
        }
    }

    fun alignLeft(otherView: View, marginDp: Float? = null, margin: Float? = null): PositionAnimExpectation {
        return PositionAnimExpectationAlignLeft(otherView).apply {
            expectations.add(this)
            this.marginDp = marginDp
            this.margin = margin
        }
    }

    fun alignRight(otherView: View, marginDp: Float? = null, margin: Float? = null): PositionAnimExpectation {
        return PositionAnimExpectationAlignRight(otherView).apply {
            expectations.add(this)
            this.marginDp = marginDp
            this.margin = margin
        }
    }

    fun originalScale(): ScaleAnimExpectation {
        return ScaleAnimExpectationOriginalScale().apply {
            expectations.add(this)
        }
    }

    fun scale(scaleX: Float, scaleY: Float, @GravityScaleHorizontalIntDef gravityHorizontal: Int, @GravityScaleVerticalIntDef gravityVertical: Int): ScaleAnimExpectation {
        return ScaleAnimExpectationValues(scaleX, scaleY, gravityHorizontal, gravityVertical).apply {
            expectations.add(this)
        }
    }

    fun scale(scaleX: Float, scaleY: Float): ScaleAnimExpectation {
        return ScaleAnimExpectationValues(scaleX, scaleY, null, null).apply {
            expectations.add(this)
        }
    }

    fun height(height: Int, @GravityScaleHorizontalIntDef horizontalGravity: Int? = null, @GravityScaleVerticalIntDef verticalGravity: Int? = null, keepRatio: Boolean = false, toDp: Boolean = false): ScaleAnimExpectation {
        return ScaleAnimExpectationHeight(height, horizontalGravity, verticalGravity).apply {
            this.keepRatio = keepRatio
            this.toDp = toDp
            expectations.add(this)
        }
    }

    fun width(width: Int, @GravityScaleHorizontalIntDef horizontalGravity: Int? = null, @GravityScaleVerticalIntDef verticalGravity: Int? = null, keepRatio: Boolean = false, toDp: Boolean = false): ScaleAnimExpectation {
        return ScaleAnimExpectationWidth(width, horizontalGravity, verticalGravity).apply {
            expectations.add(this)
            this.keepRatio = keepRatio
            this.toDp = toDp
        }
    }

    fun sameScaleAs(otherView: View): ScaleAnimExpectation {
        return ScaleAnimExpectationSameScaleAs(otherView).apply {
            expectations.add(this)
        }
    }

    fun sameWidthAs(otherView: View): ScaleAnimExpectation {
        return ScaleAnimExpectationSameWidthAs(otherView, null, null).apply {
            expectations.add(this)
        }
    }

    fun sameHeightAs(otherView: View): ScaleAnimExpectation {
        return ScaleAnimExpectationSameHeightAs(otherView, null, null).apply {
            expectations.add(this)
        }
    }

    fun toHaveAlpha(alpha: Float): AlphaAnimExpectation {
        return AlphaAnimExpectationValue(alpha).apply {
            expectations.add(this)
        }
    }

    fun sameAlphaAs(otherView: View): AlphaAnimExpectation {
        return AlphaAnimExpectationValue(otherView.alpha).apply {
            expectations.add(this)
        }
    }

    fun visible(): AlphaAnimExpectation {
        return AlphaAnimExpectationValue(1f).apply {
            expectations.add(this)
        }
    }

    fun invisible(): AlphaAnimExpectation {
        return AlphaAnimExpectationValue(0f).apply {
            expectations.add(this)
        }
    }

    fun textColor(textColor: Int): CustomAnimExpectation {
        return TextColorAnimExpectation(textColor).apply {
            expectations.add(this)
        }
    }

    fun backgroundAlpha(alpha: Float): CustomAnimExpectation {
        return ViewBackgroundAlphaAnimExpectation(alpha).apply {
            expectations.add(this)
        }
    }


    fun toBeRotated(rotation: Float): RotationExpectation {
        return RotationExpectationValue(rotation).apply {
            expectations.add(this)
        }
    }

    fun flippedHorizontally(): RotationExpectation {
        return RotationFlipExpectationValue(0f, 180f).apply {
            expectations.add(this)
        }
    }

    fun flippedVertically(): RotationExpectation {
        return RotationFlipExpectationValue(180f, 0f).apply {
            expectations.add(this)
        }
    }

    fun flippedHorizontallyAndVertically(): RotationExpectation {
        return RotationFlipExpectationValue(180f, 180f).apply {
            expectations.add(this)
        }
    }

    fun vertical(bottomOfViewAtLeft: Boolean): RotationExpectation {
        val expect: RotationExpectation
        if (bottomOfViewAtLeft) {
            expect = RotationExpectationValue(90f)
        } else {
            expect = RotationExpectationValue(270f)
        }
        expectations.add(expect)
        return expect
    }

    fun originalRotation(): RotationExpectation {
        return RotationExpectationOriginal().apply {
            expectations.add(this)
        }
    }
}
